### ISSUE TITLE
Export data as CSV

### ISSUE THREAD

**Issue by brrd** [hql287/Manta#96]
**Is it a bug or a feature request?**

- [ ] It's a bug.
- [x] I'm requesting a new feature.

**Which version are you using?**

- [ ] macOS
- [x] Windows
- [ ] Linux

**Which is the version/distribution of the OS?**

Win10 64 bits

**Describe your issue**

First of all, thank for this very promising app.

IMO adding a way to export invoices data in a human-readable format (such as CSV) would make the app more reliable for users, who don't want to loose their data if the app is discontinued one day.

Adding a way to perform this export automatically (each time an invoice is saved, for example), or with a command which could be included in a cron job, would be even better, since it would allow to backup data periodically.

**Comment by hql287** [comment]
This is a great idea! Thanks!

**Comment by mathieudutour** [comment]
yeah, while I really like having control of my data, it would be nice to be able to sync the invoices through dropbox or similar

**Comment by AlphaStyle** [comment]
Im new to contributing and have never worked with Electron before so please be gentle :smile: 

I tried to look into this, using [pouchdb-replication-stream](https://github.com/pouchdb-community/pouchdb-replication-stream) and [node-memorystream](https://github.com/JSBizon/node-memorystream) to convert the PouchDB to JSON in memory before writing to a file (Export / backup).

Im using `dialog.showOpenDialog(...)` and `dialog.showSaveDialog(...)` for opening and saving files.

There are two pouchDB's (`contacts` and `invoices`) and I think it is more user friendly if we merge these into one export file rather than two so I merge both into one JSON object in memory before I use [csvjson](https://github.com/pradeep-mishra/csvjson) to convert `JSON to CSV` and write it to a file. *(This is working so far)*

The Problem occured when I want to import this exported CSV file because I must first convert it to JSON. When using [csvjson](https://github.com/pradeep-mishra/csvjson) the input JSON and output JSON are not the same :disappointed: 

So **the question**: Do you know of any `JSON to CSV` - `CSV to JSON` packages? or should we write our own script?

**Comment by hql287** [comment]
@AlphaStyle Hey, sorry for the delayed response. Any updates on this so far?

**Comment by AlphaStyle** [comment]
I have to admit I haven't done much since last time I wrote.
I was hoping someone could provide me with this magical package that will fix my `JSON->CSV | CSV->JSON` issues.

As of right now I can export everything to one single CSV file. (The import logic is there, just not working cause of `CSV->JSON` is not correct).

I guess I could try and write my own `CSV-JSON` converter, but I'm not the most experienced coder :smile: But could be good practice though :+1:

**Comment by hql287** [comment]
So `JSON` to `CSV` is fine but you need some help with `CSV` to `JSON`?

**Comment by hql287** [comment]
> I guess I could try and write my own CSV-JSON converter, but I'm not the most experienced coder 😄 But could be good practice though 👍

Definitely a good exercise, and you'll help lots of other fellow developers with it 👍

**Comment by AlphaStyle** [comment]
Yes, `CSV` to `JSON`. The PouchDB JSON dump is what I think people refer to as complex nested object with arrays. So [csvjson](https://github.com/pradeep-mishra/csvjson) package does something wrong when it covert CSV to JSON.

Maybe I could try and contribute to [csvjson](https://github.com/pradeep-mishra/csvjson) instead of starting from scratch.

**Comment by brrd** [comment]
@AlphaStyle What about https://www.npmjs.com/package/csv-to-pouch module?

**Comment by hql287** [comment]
But you'll only need `CSV` to `JSON` for importing data, correct? If yes, how about making this 2 separated PRs, one for export, one for import. I can use the export module for https://github.com/hql287/Manta/issues/124

**Comment by AlphaStyle** [comment]
Unfortunately [csv-to-pouch](https://www.npmjs.com/package/csv-to-pouch) won't work for my approach because Manta uses two PouchDB's (Invoices and Contacts) and **csv-to-pouch** does not return the JSON, only add it directly to PouchDB *(please correct me if I am wrong)*. 

I am hoping we can continue to have only one exported CSV file to maintain the simplicity for the end-user.

Maybe I should provide more information / pictures of my issue so people can know exactly what I am strugling with :smile: 

I might have it all wrong though, maybe I am over thinking it. [PouchDB dump](https://github.com/pouchdb-community/pouchdb-replication-stream#design), here you can see an example of the JSON I get from PouchDB dump. In theory all I need is the **docs** objects because they are the ones with the important content, but as of right now I save EVERYTHING.

**Comment by AlphaStyle** [comment]
@hql287 Ok, let me just double check it works and then I will PR it :)

**Comment by mathieudutour** [comment]
oh dumping the entire pouchdb probably won't work: a csv is more or less an array of JSON objects.

So what you really want is only the docs. We don't really about all the metadata (I would even remove the `_rev` from the docs).

Then the `csvjson` will probably work

**Comment by hql287** [comment]
> So what you really want is only the docs 

👍 This is what I would do, too.

**Comment by AlphaStyle** [comment]
Im only storing the **docs** now. Still not working the way I want it :/
I updated everything to fit in with the recent changes since last time I developed.

Do I need permission to make a pull request or am I doing something realy wrong?

**Comment by mathieudutour** [comment]
if you have a fork, no.

What does your JSON look like?

**Comment by AlphaStyle** [comment]
I really wanna do a pull request so you guys can see for yourself :smile: 

*ERROR: Permission to hql287/Manta.git denied to Alphastyle.
Please make sure you have the correct access rights and the repository exists*

I have never experienced this :disappointed:

**Comment by hql287** [comment]
> I have never experienced this 😞

Well, then you're in luck today, haha.

Kidding aside, here's how you would do it:
* Create a fork of this repo
* Work on the PR locally on your machine
* Back to this repo home page and [create a new PR](https://github.com/hql287/Manta/compare)
* Click on the _compare across forks_ 
![compare across forks ](https://d2ffutrenqvap3.cloudfront.net/items/2v3n1K1v0M43003V0V15/Screen%20Shot%202018-01-05%20at%2012.52.35.png?v=ab92001a)
* The base fork should be this repo dev branch and the head fork should be the PR branch on your fork.
* GitHub will diff the 2 branches and let you know if it's possible to create the PR

**Comment by hql287** [comment]
GitHub also has a detailed guide on [how to create a PR from a fork](https://help.github.com/articles/creating-a-pull-request-from-a-fork/).

**Comment by hql287** [comment]
> I have never experienced this 😞

Don't worry, we all started somewhere. 
GitHub also has a detailed guide on [how to create a PR from a fork](https://help.github.com/articles/creating-a-pull-request-from-a-fork/).

**Comment by AlphaStyle** [comment]
I am sorry for super late reply!
I do appreciate your patient and wisdom :smile: 

I did not know about the fork approach. I will read the guide and make a PR :smile:

**Comment by jens-t** [comment]
I think @mathieudutour has a point. CSV is ok for human readability, and its fine to export it with a click on a button. But additionally some realtime export to a folder set in the settings of the app, would be what i'm more interested in. So that you can watch the folder for filechanges with whatever syncinc tool you'r using :)

Opinions?

**Comment by hql287** [comment]
> But additionally some realtime json export to a folder set in the settings of the app, would be what i'm more interested in. So that you can watch the folder for filechanges with whatever syncing tool you'r using :)

Sounds interesting. One of the reasons that I chose `pouchDB` was because of its ability to sync with a remote server. But this should be a separate feature and deserve its own PR. 

@AlphaStyle has already been working on this for some time. You might want to take a look here: https://github.com/hql287/Manta/pull/143.

**Comment by jens-t** [comment]
> But this should be a separate feature and deserve its own PR.

Absolutely! Just wanted some feedback before starting a new feature request :)

**Comment by hql287** [comment]
Sure, will get back to this after I'm done with releasing a v1.1.3
